sdk/core/azure-core/src/http/websockets/websockets_impl.cpp

```diff
@@ -0,0 +1,884 @@
+          context);
+    }
+    else
+#endif
+    {
+      //      Log::Write(Logger::Level::Verbose, "Send Binary Frame " + HexEncode(binaryFrame, 16));
+      std::vector<uint8_t> sendFrame
+          = EncodeFrame(SocketOpcode::BinaryFrame, isFinalFrame, binaryFrame);
+
+      SendTransportBuffer(sendFrame, context);
+    }
+  }
+
+  std::shared_ptr<WebSocketFrame> WebSocketImplementation::ReceiveFrame(
+      Azure::Core::Context const& context)
+  {
+    std::unique_lock<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+
+    if (m_state != SocketState::Open && m_state != SocketState::Closing)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+
+    // Unlock the state lock to allow other threads to run. If we don't, we might end up in in a
+    // situation where the server won't respond to the this client because all the client threads
+    // are blocked on the state lock.
+    lock.unlock();
+
+    std::shared_ptr<WebSocketInternalFrame> frame;
+    // Loop until we receive an returnable incoming frame.
+    // If the incoming frame is returnable, we return the value from the frame.
+    while (true)
+    {
+      frame = ReceiveTransportFrame(context);
+      if (frame)
+      {
+        switch (frame->Opcode)
+        {
+            // When we receive a "ping" frame, we want to send a Pong frame back to the server.
+          case SocketOpcode::Ping:
+            Log::Write(
+                Logger::Level::Verbose, "Received Ping frame: " + HexEncode(frame->Payload, 16));
+            SendPong(frame->Payload, context);
+            break;
+            // We want to ignore all incoming "Pong" frames.
+          case SocketOpcode::Pong:
+            Log::Write(
+                Logger::Level::Verbose, "Received Pong frame: " + HexEncode(frame->Payload, 16));
+            break;
+
+          case SocketOpcode::BinaryFrame:
+            m_currentMessageType = SocketMessageType::Binary;
+            return std::shared_ptr<WebSocketFrame>(new WebSocketBinaryFrame(
+                frame->IsFinalFrame, frame->Payload.data(), frame->Payload.size()));
+
+          case SocketOpcode::TextFrame:
+            m_currentMessageType = SocketMessageType::Text;
+            return std::shared_ptr<WebSocketFrame>(new WebSocketTextFrame(
+                frame->IsFinalFrame, frame->Payload.data(), frame->Payload.size()));
+
+          case SocketOpcode::Close: {
+            if (frame->Payload.size() < 2)
+            {
+              throw std::runtime_error("Close response buffer is too short.");
+            }
+            uint16_t errorCode = 0;
+            errorCode |= (frame->Payload[0] << 8) & 0xff00;
+            errorCode |= (frame->Payload[1] & 0x00ff);
+
+            // We received a close frame, mark the socket as closed. Make sure we
+            // reacquire the state lock before setting the state to closed.
+            lock.lock();
+            m_state = SocketState::Closed;
```

Comment:
Would we benefit by protecting the m_state in a method that ensures the lock is held before modiying?  So instead of the lock mutex acquistion in each method you call ChangeState(newState) which ensures the lock gets taken and released around the state change.

Reply:
Maybe this isn't as clean as it appears.  There are probably places you want to hold the state lock for longer???

Reply:
It's entirely possible that there are logic errors here. Let me think about it a bit...